pkg/controllers/jobset_controller.go

```diff
@@ -916,3 +939,98 @@ func findReplicatedStatus(replicatedJobStatus []jobset.ReplicatedJobStatus, repl
 	}
 	return jobset.ReplicatedJobStatus{}
 }
+
+func (r *JobSetReconciler) deleteJobSet(ctx context.Context, js *jobset.JobSet) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	policy := metav1.DeletePropagationForeground
+	options := []client.DeleteOption{client.PropagationPolicy(policy)}
+	log.V(2).Info("Cleaning up JobSet", "jobset", klog.KObj(js))
+
+	return r.Delete(ctx, js, options...)
+}
+
+// checkIfTTLExpired checks whether a given JobSet's TTL has expired.
+func (r *JobSetReconciler) checkIfTTLExpired(ctx context.Context, jobSet *jobset.JobSet) (*time.Time, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// We don't care about the JobSets that are going to be deleted
+	if jobSet.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	now := r.clock.Now()
+	remaining, expiredAt, err := timeLeft(log, jobSet, &now)
+	if err != nil {
+		return nil, err
+	}
+
+	// TTL has expired
+	if remaining != nil && *remaining <= 0 {
+		return expiredAt, nil
+	}
+
+	return nil, nil
+}
+
+// timeLeft returns the time left until the JobSet's TTL expires and the time when it will expire.
+func timeLeft(log logr.Logger, js *jobset.JobSet, since *time.Time) (*time.Duration, *time.Time, error) {
+	finishAt, expireAt, err := getJobSetFinishAndExpireTime(js)
+	if err != nil {
+		return nil, nil, err
+	}
+	// The following 2 checks do sanity checking for nil pointers in case of changes to the above function.
+	// This logic should never be executed.
+	if since == nil {
+		log.V(2).Info("Warning: Received invalid since time. JobSet cleanup will be deferred.")
+		return nil, nil, nil
+	}
+	if finishAt == nil || expireAt == nil {
+		log.V(2).Info("Warning: Received invalid finish or expire time. JobSet cleanup will be deferred.")
```

Comment:
I think these 3 nil checks be combined into 1 right? Any reason you kept them separate? If not let's combine them.

Reply:
We can, I was wondering should have different logs for each case, but that might be overkill for a scenario which we won't ever encounter most probably